File: python/dijkstra_gui_tk/undirected_graph.py
import random

"""
An UndirectedGraph has bi-directional traversal paths between nodes.
"""
class UndirectedGraph:
  def __init__(self):
    self.paths = {}

  def addPath(self, fromNode, toNode, cost):
    if not fromNode in self.paths:
      self.paths[fromNode] = {}
    self.paths[fromNode][toNode] = cost

  def neighborsOf(self, node):
    if not node in self.paths:
      return {}
    return self.paths[node]

  def randomNode(self):
    return random.choice(list(self.paths))

File: ruby/dijkstra/undirected_graph.rb
class UndirectedGraph

  attr_reader :paths

  def initialize
    @paths = {}
  end

  def add_path(from, to, cost)
    get_paths(from)[to] = cost
  end

  def get_paths(from)
    @paths[from] ||= {}
  end

  def get_random_path
    return nil unless @paths.keys.size > 0

    get_paths(@paths.keys.sample)
  end

end
